v3_usm.go

```diff
@@ -112,32 +112,33 @@ func (sp *UsmSecurityParameters) setSecurityParameters(in SnmpV3SecurityParamete
 
 	if sp.AuthoritativeEngineID != insp.AuthoritativeEngineID {
 		sp.AuthoritativeEngineID = insp.AuthoritativeEngineID
-	}
-	if sp.AuthenticationProtocol > NoAuth && len(sp.SecretKey) == 0 {
-		sp.SecretKey, err = genlocalkey(sp.AuthenticationProtocol,
-			sp.AuthenticationPassphrase,
-			sp.AuthoritativeEngineID)
-		if err != nil {
-			return err
-		}
-	}
-	if sp.PrivacyProtocol > NoPriv && len(sp.PrivacyKey) == 0 {
-		switch sp.PrivacyProtocol {
-		// Changed: The Output of SHA1 is a 20 octets array, therefore for AES128 (16 octets) either key extension algorithm can be used.
-		case AES, AES192, AES256, AES192C, AES256C:
-			//Use abstract AES key localization algorithms
-			sp.PrivacyKey, err = genlocalPrivKey(sp.PrivacyProtocol, sp.AuthenticationProtocol,
-				sp.PrivacyPassphrase,
+
+		if sp.AuthenticationProtocol > NoAuth && len(sp.SecretKey) == 0 {
+			sp.SecretKey, err = genlocalkey(sp.AuthenticationProtocol,
+				sp.AuthenticationPassphrase,
 				sp.AuthoritativeEngineID)
 			if err != nil {
 				return err
 			}
-		default:
-			sp.PrivacyKey, err = genlocalkey(sp.AuthenticationProtocol,
-				sp.PrivacyPassphrase,
-				sp.AuthoritativeEngineID)
-			if err != nil {
-				return err
+		}
+		if sp.PrivacyProtocol > NoPriv && len(sp.PrivacyKey) == 0 {
+			switch sp.PrivacyProtocol {
+			// Changed: The Output of SHA1 is a 20 octets array, therefore for AES128 (16 octets) either key extension algorithm can be used.
+			case AES, AES192, AES256, AES192C, AES256C:
+				//Use abstract AES key localization algorithms
+				sp.PrivacyKey, err = genlocalPrivKey(sp.PrivacyProtocol, sp.AuthenticationProtocol,
+					sp.PrivacyPassphrase,
+					sp.AuthoritativeEngineID)
+				if err != nil {
+					return err
+				}
+			default:
+				sp.PrivacyKey, err = genlocalkey(sp.AuthenticationProtocol,
+					sp.PrivacyPassphrase,
+					sp.AuthoritativeEngineID)
+				if err != nil {
+					return err
+				}
```

Comment:
This was reverted to the original form before my bugfix, the actual bugfix is in gosnmp.go and can be seen here:
https://github.com/soniah/gosnmp/commit/715e266c7913d7765af2e6da61d65c9dfce4839e

Reply:
Thanks very much guys! I'll have a look at this over the weekend